### PR TITLE
Progress Tracker APIS

### PR DESCRIPTION
**Pull request by priyanshu20**
Group CRUD
Task CRUD
Task Assignee

### CODE REVIEW

File: models/TaskAssignee.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const mongoose = require("mongoose");
+
+const TaskAssigneeSchema = new mongoose.Schema(
+	{
+		assigneeId: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
+		assignedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
+		status: {
+			type: String,
+			default: "pending",
+			enum: ["done", "onit", "pending", "overdue"]
```
Comment (rhnmht30) [inline models/TaskAssignee.js:10]: can we have these options for tasks status?
`[
 "pending",
 "ongoing",
 "completed",
 "closed",
 "overdue"
]`

they look more modular to me

File: models/TaskAssignee.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+const mongoose = require("mongoose");
+
+const TaskAssigneeSchema = new mongoose.Schema(
+	{
+		assigneeId: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
+		assignedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
```
Comment (rhnmht30) [inline models/TaskAssignee.js:6]: IMO I don't see the need for `assignedBy`. We have that ref in the task schema. A task is created by the group head and this field will not change when the same task is assigned to several people.

File: controllers/group_controller.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		}
+	]);
+	return sendSuccess(res, groups);
+};
+
+module.exports.addGroup = async (req, res) => {
+	let { name, heads, members } = req.body;
+	heads = [...new Set(heads)];
+	members = [...new Set(members)];
+	if ((await User.count({ _id: { $in: heads } })) !== heads.length)
```
Comment (rhnmht30) [inline controllers/group_controller.js:49]: Can we query this count in a single query? @ritiksr25

Comment (rhnmht30) [reply controllers/group_controller.js:49]: Maybe create a common set of both _id and then query.

File: controllers/task_controller.js [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+		$match: {}
+	};
+	let group = await Group.findById(gid);
+	if (!group) return sendError(res, "Group not found!!", NOT_FOUND);
+
+	if (
+		!group.heads.includes(ObjectId(req.user.id)) &&
+		!["lead", "core"].includes(req.user.role)
+	)
+		return sendError(res, "Not allowed to do that!!", FORBIDDEN);
```
Comment (rhnmht30) [inline controllers/task_controller.js:31]: Be more clear on the error messages

File: routes/api/v1/tasks.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+const { allAuth, coreAuth } = require("../../../middlewares/auth");
+const {
+	addTaskValidation,
+	updateTaskValidation,
+	updateTaskAssigneeValidation
+} = require("../../../middlewares/validations");
+
+router.get(
+	"/assignees/:taid",
+	catchErrors(coreAuth),
```
Comment (rhnmht30) [inline routes/api/v1/tasks.js:26]: This won't a `member` who is head of a group to get the list of assignees.

File: controllers/task_controller.js [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+			$lookup: {
+				from: "tasks",
+				localField: "_id",
+				foreignField: "taskAssignees",
+				as: "tasksData"
+			}
+		},
+		{
+			$facet: {
+				"Assigned Tasks": [
```
Comment (rhnmht30) [inline controllers/task_controller.js:78]: Are these tasks assigned by the user or to the user?

File: routes/api/v1/tasks.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	"/assignees/:taid",
+	catchErrors(allAuth),
+	catchErrors(updateTaskAssigneeValidation),
+	catchErrors(updateTaskAssignee)
+);
+
+// FETCH TASKS
+router.get("/:gid", catchErrors(allAuth), catchErrors(getTasks));
+router.get("/:gid/my", catchErrors(allAuth), catchErrors(getMyTasks));
+router.get("/:gid/:uid", catchErrors(coreAuth), catchErrors(getUserTasks));
```
Comment (rhnmht30) [inline routes/api/v1/tasks.js:40]: This won't allow a member who is head of a group to get the list of user tasks for their group.

**Review by rhnmht30 — CHANGES_REQUESTED** [review]
We need to set up a meet to discuss the flow once again.

File: models/TaskAssignee.js (hunk 1 above)
Comment (priyanshu20) [reply models/TaskAssignee.js:10]: Yes, they seem better

File: controllers/task_controller.js (hunk 6 above)
Comment (priyanshu20) [reply controllers/task_controller.js:78]: Assigned to the user.

File: routes/api/v1/tasks.js (hunk 5 above)
Comment (priyanshu20) [reply routes/api/v1/tasks.js:26]: this is specifically for core/lead, the usual workflow won't require fetching taskAssignees  through an API, as they will be provided when fetching the task anyways.

only updating taskassignees is required as a separate route, which is open to members

File: controllers/group_controller.js (hunk 3 above)
Comment (priyanshu20) [reply controllers/group_controller.js:49]: Yep , that will save an extra query 👍🏽

File: models/TaskAssignee.js (hunk 2 above)
Comment (priyanshu20) [reply models/TaskAssignee.js:6]: When fetching tasks for a user, assignedBy is used to fetch tasks created by a user whereas assigneeId is used to fetch tasks assigned to the user

Comment (rhnmht30) [reply models/TaskAssignee.js:6]: okay cool!

File: controllers/todo_controller.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
+const Todo = require("../models/Todo");
```
Comment (ritiksr25) [inline controllers/todo_controller.js:1]: todos_controller is already there, rename or remove this file:)

**Review by ritiksr25 — CHANGES_REQUESTED** [review]
naice

File: controllers/group_controller.js [hunk 9]
```diff
@@ -11,45 +11,180 @@ const { sendError, sendSuccess } = require("../utility/helpers");
...
+								pipeline: [
+									{
+										$match: {
+											$expr: {
+												$in: ["$_id", "$$member_list"]
+											}
+										}
+									},
+									{
+										$sort: { lastLogin: -1 }
```
Comment (ritiksr25) [inline controllers/group_controller.js:79]:
```suggestion
										$sort: { lastActiveAt: -1 }
```

File: controllers/group_controller.js [hunk 10]
```diff
@@ -11,45 +11,180 @@ const { sendError, sendSuccess } = require("../utility/helpers");
...
```
Comment (ritiksr25) [inline controllers/group_controller.js:43]:
```suggestion
										$sort: { lastActiveAt: -1 }
```

File: controllers/group_controller.js [hunk 11]
```diff
@@ -11,45 +11,180 @@ const { sendError, sendSuccess } = require("../utility/helpers");
...
```
Comment (ritiksr25) [inline controllers/group_controller.js:120]:
```suggestion
										$sort: { lastActiveAt: -1 }
```

File: controllers/group_controller.js [hunk 12]
```diff
@@ -11,45 +11,180 @@ const { sendError, sendSuccess } = require("../utility/helpers");
...
```
Comment (ritiksr25) [inline controllers/group_controller.js:156]:
```suggestion
										$sort: { lastActiveAt: -1 }
```

**Review by ritiksr25 — APPROVED** [review]
naice

**Comment by rhnmht30** [comment]
damnn! nice 🎉

**Pull request merged**